src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -0,0 +1,382 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.memory;
+
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.SortedSet;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.util.concurrent.FastThreadLocal;
+import org.apache.cassandra.db.Clustering;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.memtable.TrieMemtable;
+import org.apache.cassandra.db.tries.InMemoryTrie;
+import org.apache.cassandra.db.tries.Trie;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.analyzer.AbstractAnalyzer;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.PrimaryKeys;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+
+/**
+ * This is an in-memory index using the {@link InMemoryTrie} to store a {@link ByteComparable}
+ * representation of the indexed values. Data is stored on-heap or off-heap and follows the
+ * settings of the {@link TrieMemtable} to determine where.
+ *
+ *
+ */
+public class TrieMemoryIndex
+{
+    private static final Logger logger = LoggerFactory.getLogger(TrieMemoryIndex.class);
+    private static final int MINIMUM_QUEUE_SIZE = 128;
+    private static final int MAX_RECURSIVE_KEY_LENGTH = 128;
+
+    private final IndexContext indexContext;
+    private final InMemoryTrie<PrimaryKeys> data;
+    private final PrimaryKeysReducer primaryKeysReducer;
+    private final AbstractAnalyzer.AnalyzerFactory analyzerFactory;
+    private final AbstractType<?> validator;
+    private final boolean isLiteral;
+    private final Object writeLock = new Object();
+
+    private ByteBuffer minTerm;
+    private ByteBuffer maxTerm;
+
+    private static final FastThreadLocal<Integer> lastQueueSize = new FastThreadLocal<Integer>()
```

Comment:
nit: Perhaps a quick comment on why this is useful. I'm guessing it's because the size of the last result is predictive of the size of the next one?

Reply:
Agreed. I have moved this and the associated constant into the `Collector` because that is where it is used. I've added a comment to as well as a TODO about possibly changing to using a decaying histogram (or similar) to avoid the effect of outliers. That can be another follow up ticket.